gunicorn/_reloader.py

```diff
@@ -0,0 +1,58 @@
+# -*- coding: utf-8 -
+#
+# This file is part of gunicorn released under the MIT license.
+# See the NOTICE for more information.
+
+import os.path
+import re
+import sys
+
+from watchdog.events import FileSystemEventHandler
+from watchdog.observers import Observer
+from watchdog.observers.polling import PollingObserver
+
+
+class Reloader(object):
+
+    def __init__(self, extra_files=None, interval=1, callback=None,
+                 engine='auto'):
+        super(Reloader, self).__init__()
+        if engine == 'auto' or engine == 'inotify':
+            self._observer = Observer()
+        else:
+            assert engine == 'poll'
```

Comment:
I left it as an assert because it should be caught be the configuration system, but I can change it if it still doesn't seem right to you.

Reply:
Indeed, a few people do use `-O` in production.

Reply:
@tilgovi "it should be caught by the configuration system" what do you mean?

Reply:
I looked at the issue list of the project and it seems it doesn't work everywhere. Also it doesn't sound very active. Before taking any decision i would check i it really works multi-platform as some issues let me think it is not (like on docker or osx). 

Anyway i'm not inclined to make such thing an hard dependency. Though it would be awesome to handle local reloads and not only for local developments. Silly idea but couldn't have equivalent or fork maintained directly inside gunicorn?

Reply:
It looks like https://github.com/gorakhargosh/watchdog is the repo.

It does seem that the last commits are last November, but the issues that people seem to open recently don't seem too urgent. Although, I see some good looking PRs that are stale. I would rather try to get in touch with the maintainer(s) and see if we can unblock things than to fork it into Gunicorn.

It does claim to support all these platforms, and I believe it. At the very least, it supports a fallback that works like ours did. On every platform except BSD, it seems, there's fallbacks to the polling observer: https://github.com/gorakhargosh/watchdog/blob/master/src/watchdog/observers/__init__.py

Reply:
Is your objection to a hard dependency about code size or security or bugs or... we can code defensively and make the PollingObserver the default if that feels safer.

Reply:
I've e-mailed the most recent maintainers to ask the status of the project.

Reply:
@tilgovi did you have a response?

For me automatic reload support should be a core feature of Gunicorn (like HUP is). We should even improve it by offering a good support of gunicorn for the editors and IDEs but also offere automatic reload support for prod like the one you find in caddy.

 **If we consider it is a core feature and useful for our user, then we should be able to support it.** 

I'm quite doubtful we could offer a good support of it by using a dependency. We already have some issues with some of the dependency inside gunicorn even indirect that need to be fixed. 

That's why I proposed to maybe fork it and use our own package/ internal source code (could we use submodules and still patch upstream?). In between we could a good default (internal obsever) and allow the use of watchdog if available (and maybe encourage with a log) like you propose.  Since we are going to full python 3 soon, support of kqueue and other should be easier.

Thoughts?

Reply:
I think you're right. I have not heard back from my e-mail to the watchdog maintainers. We do need this to work. We should vendor it. Maybe @kennethreitz has good thoughts because I know there is vendoring in pipenv. Rather than just copy modules into the gunicorn package, we should copy a whole package as much as possible and fork. Maybe this is a good use of the gunicorn github org?